musmart/melody/contour/step_contour.py

```
@@ -0,0 +1,161 @@
+        -------
+        list[int]
+            List of length step_contour_length containing repeated pitch values
+        """
+        result = []
+        for pitch, duration in zip(pitches, normalized_durations):
+            repetitions = round(duration)
+            result.extend([pitch] * repetitions)
+
+        if len(result) > self._step_contour_length:
+            result = result[:self._step_contour_length]
+        elif len(result) < self._step_contour_length:
+            result.extend([result[-1]] * (self._step_contour_length - len(result)))
+
+        return result
+
+    def _calculate_contour(
+        self,
+        pitches: list[int],
+        durations: list[float]
+    ) -> list[int]:
+        """Calculate the step contour from input pitches and durations."""
+        normalized_durations = self._normalize_durations(durations)
+        return self._expand_to_vector(pitches, normalized_durations)
+
+    @property
+    def contour(self) -> list[int]:
+        """Get the step contour vector."""
+        return self._contour
+
+    @property
+    def global_variation(self) -> float:
+        """Calculate the global variation of the step contour.
+
+        Returns
+        -------
+        float
+            Float value representing the global variation of the step contour
+        """
+        return float(np.nanstd(self._contour))
+
+    @property
+    def global_direction(self) -> float:
+        """Calculate the global direction of the step contour.
+
+        Returns
+        -------
+        float
+            Float value representing the global direction of the step contour
+            Returns 0.0 if the contour is flat
+        """
+        if len(set(self._contour)) == 1:
+            return 0.0
+
+        try:
+            corr = np.corrcoef(
+                self._contour,
+                np.arange(self._step_contour_length)
+            )[0, 1]
+            return float(corr)
+        except Exception:
+            return None
```

CC'ing @pmcharrison. I'm not sure about this from an API perspective. I feel fairly strongly that if we have an error, this shouldn't be skipped silently.

I understand the returning 0.0 for a flat array and think this makes sense intuitively. But I would allow `np.corrcoef` to fail rather than returning None here.

---

In a use case scenario, if I'm running this class as part of a pipeline where only one input will fail, it's possible to imagine how I could get weird errors from e.g. trying to perform further `numpy` operations on an array that contains both `floats` and `NoneType` values. This would waste time, when really the root of the problem is the operation failing.

---

Totally agree with @HuwCheston, never write `except Exception` unless you then reraise it later. Your linter should catch it. @Dav8Circle if you don't have it already, can you install the Pylint extension on your Cursor?

---

My mistake, I get you now - definitely better to just let `np.corrcoef` fail than try and catch in a confusing way. @pmcharrison I've installed Pylint as requested, hadn't set it up in this IDE!